### PR TITLE
Select toggle widget

### PR DESCRIPTION
**Pull request by orangejenny**
This pattern is in the style guide, but we don't have a good easy way to implement it.

code buddy @millerdev 
knockout opinions @proteusvacuum 

screen shot 2019-02-25 at 10 10 56 am

fyi @pr33thi - this doesn't support multiple selections yet, but once it does it'll be able to replace the submission type filter you've been fixing up recently on the raw forms, errors & duplicates report.

### CODE REVIEW

File: corehq/apps/hqwebapp/static/hqwebapp/js/components/select_toggle.js [hunk 1]
```diff
@@ -0,0 +1,54 @@
+/*
+ * Component for displaying a selection as a set of buttons.
+ * Creates an invisible <select> element to store the value and a set of buttons for the user to interact with.
+ *
+ * Required parameters
+ *  - options: Data to build select options. May be list of strings or list of objects with `id` and `text` properties.
```
Comment (orangejenny) [inline corehq/apps/hqwebapp/static/hqwebapp/js/components/select_toggle.js]: It likely makes sense for this to also accept an observable (same thing for the value param). I was lazy for now.

File: corehq/apps/hqwebapp/static/hqwebapp/js/components/select_toggle.js [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+                return {
+                    id: id,
+                    text: text,
+                    selected: ko.computed(function () {
+                        return id === self.value();
+                    }),
+                    click: function (model, e) {
+                        if (model.id !== self.value()) {
+                            self.value(model.id);
+                            $(e.currentTarget).closest(".ko-select-toggle").find("select").trigger("change");
```
Comment (orangejenny) [inline corehq/apps/hqwebapp/static/hqwebapp/js/components/select_toggle.js]: Very un-knockout-y behavior here, did some experimenting with custom bindings but didn't find a better way to do it.

**Comment by orangejenny** [comment]
fyi @dimagi/product

**Review by biyeun — APPROVED** [review]
looks solid. thanks, jenny! :fire:

File: corehq/apps/hqwebapp/static/hqwebapp/js/components/select_toggle.js [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+        viewModel: function (params) {
+            var self = this;
+
+            // Attributes passed on to the input
+            self.name = params.name || '';
+            self.id = params.id || '';
+
+            // Data
+            self.value = ko.observable(params.value || '');
+            self.options = ko.observableArray(_.map(params.options, function (o) {
```
Comment (millerdev) [inline corehq/apps/hqwebapp/static/hqwebapp/js/components/select_toggle.js]: 👀 single letter variable

Comment (orangejenny) [reply corehq/apps/hqwebapp/static/hqwebapp/js/components/select_toggle.js]: eh, I don't mind them in one-statement callbacks like this

Comment (millerdev) [reply corehq/apps/hqwebapp/static/hqwebapp/js/components/select_toggle.js]: You call _that_ (function spanning lines 33-50) a one-statement callback? Granted, this variable is only used on the first three lines, but I had to scan the whole function to figure that out.

Comment (orangejenny) [reply corehq/apps/hqwebapp/static/hqwebapp/js/components/select_toggle.js]: I responded to this comment based on the email, which only showed up to this line. Forgot how long this callback is. I don't love how big this function got, or the fact that it contains an anonymous knockout model, but I'm going to leave it alone for now.

**Pull request merged**